lib-ui-compose/src/commonMain/kotlin/com/ustadmobile/libuicompose/view/report/graphs/Graphs.kt

```diff
@@ -0,0 +1,454 @@
+            modifier = Modifier.size(16.dp),
+            fillBrush = SolidColor(color),
+            shape = RoundedCornerShape(4.dp)
+        )
+        Spacer(modifier = Modifier.width(4.dp))
+        Text(label)
+    }
+}
+
+@Composable
+private fun AxisLabels(label: String, modifier: Modifier = Modifier) {
+    androidx.compose.material3.Text(
+        label,
+        modifier = modifier.fillMaxWidth(),
+        overflow = TextOverflow.Ellipsis,
+        maxLines = 1,
+        textAlign = androidx.compose.ui.text.style.TextAlign.Center,
+    )
+}
+
+@Composable
+private fun AxisValue(label: String, modifier: Modifier = Modifier) {
+    androidx.compose.material3.Text(
+        label,
+        modifier = modifier.fillMaxWidth(),
+        overflow = TextOverflow.Ellipsis,
+        maxLines = 1,
+        textAlign = androidx.compose.ui.text.style.TextAlign.End,
+        fontSize = 8.sp
+    )
+}
+
+@Composable
+private fun HoverSurface(content: @Composable () -> Unit) {
+    Surface(
+        shadowElevation = 4.dp,
+        shape = androidx.compose.material3.MaterialTheme.shapes.medium,
+        color = androidx.compose.material3.MaterialTheme.colorScheme.surfaceContainerHighest,
+        modifier = Modifier.defaultItemPadding(4.dp)
+    ) {
+        Box(modifier = Modifier.defaultItemPadding(8.dp)) {
+            content()
+        }
+    }
+}
+
+private fun calculateSubgroupValues(
+    barSeries: List<GraphSeries>,
+    xValue: Any,
+    subgroup: String,
+    conversionFactor: Double
+): Float {
+    return barSeries.flatMap { barSeriesItem ->
+        barSeriesItem.data.filter { dataPoint ->
+            dataPoint.xAxis == xValue && (dataPoint.subgroup ?: "") == subgroup
+        }
+    }.firstOrNull()?.yAxis?.times(conversionFactor)?.toFloat() ?: 0f
+}
+
+private fun calculateConversionFactor(isDuration: Boolean, maxY: Double): Pair<Double, String> {
+    return when {
+        isDuration -> {
+            when {
+                maxY >= 3_600_000 -> Pair(1.0 / 3_600_000, HOUR_UNIT)
+                maxY >= 60_000 -> Pair(1.0 / 60_000, MINUTE_UNIT)
+                else -> Pair(1.0 / 1_000, SECOND_UNIT)
+            }
+        }
+
+        else -> Pair(1.0, "")
+    }
+}
+
+@Composable
+fun formatSubgroupValue(
+    value: String,
+    reportOptions: ReportOptions2?,
+): String {
+    return reportOptions?.series?.firstOrNull()?.reportSeriesSubGroup?.let { axisType ->
+        when (axisType) {
+            ReportXAxis.GENDER -> getGenderLabel(value)
+            ReportXAxis.CLASS -> value
+            else -> ReportFormatter.formatDateForReport(value, axisType)
+        }
+    } ?: value
+}
+
+private fun calculateTickIncrement(
+    yRange: ClosedFloatingPointRange<Float>,
+    yAxisLabel: String,
+    unit: String,
+    isDurationType: Boolean
+): Float {
+    val range = yRange.endInclusive - yRange.start
+    return when {
+        range == 0f -> 1f
+        yAxisLabel.equals(YAxisTypes.COUNT.name, ignoreCase = true) -> {
```

Comment:
This would definitely break in any non-English language; and could break even in English if the user simply changed the label.